<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Ancestral state reconstruction — ancRECON • corHMM</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Ancestral state reconstruction — ancRECON"><meta property="og:description" content="Infers ancestral states based on a set of model parameters"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">corHMM</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.9</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/corHMMv2.1-vignette.html">corHMM 2.1: Generalized hidden Markov models</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Ancestral state reconstruction</h1>
    
    <div class="hidden name"><code>ancRECON.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Infers ancestral states based on a set of model parameters</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">ancRECON</span><span class="op">(</span><span class="va">phy</span>,<span class="va">data</span>, <span class="va">p</span>, method<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"joint"</span>, <span class="st">"marginal"</span>, <span class="st">"scaled"</span><span class="op">)</span>,</span>
<span><span class="va">rate.cat</span>, ntraits<span class="op">=</span><span class="cn">NULL</span>, rate.mat<span class="op">=</span><span class="cn">NULL</span>,</span>
<span>model<span class="op">=</span><span class="st">"ARD"</span>, root.p<span class="op">=</span><span class="cn">NULL</span>, get.likelihood<span class="op">=</span><span class="cn">FALSE</span>, get.tip.states <span class="op">=</span> <span class="cn">FALSE</span>, collapse <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>phy</dt>
<dd><p>a phylogenetic tree, in <code>ape</code> “phylo” format.</p></dd>

<dt>data</dt>
<dd><p>a data matrix containing species information (see Details).</p></dd>

<dt>p</dt>
<dd><p>a vector of transition rates to be used to estimate ancestral states.</p></dd>

<dt>method</dt>
<dd><p>method used to calculate ancestral states at internal nodes. Can be one of: "joint", "marginal", or "scaled" (see Details).</p></dd>

<dt>rate.cat</dt>
<dd><p>specifies the number of rate categories in the HRM.</p></dd>
 
<dt>ntraits</dt>
<dd><p>currently, this is automaticall detected and can always be set to NULL.</p></dd>

<dt>rate.mat</dt>
<dd><p>a user-supplied rate matrix index of parameters to be optimized.</p></dd>

<dt>model</dt>
<dd><p>specifies the underlying model if a rate.mat is not provided ("ER", SYM", or "ARD").</p></dd>

<dt>root.p</dt>
<dd><p>a vector used to fix the probabilities at the root, but “yang” and “maddfitz” can also be supplied to use the method of Yang (2006) and FitzJohn et al (2009) respectively (see details).</p></dd>

<dt>get.likelihood</dt>
<dd><p>a logical indicating whether to obtain the likelihood of the rates and states. The default is <code>FALSE</code>.</p></dd>

<dt>get.tip.states</dt>
<dd><p>a logical indicating whether just tip reconstructions should be output. The default is <code>FALSE</code>.</p></dd>

<dt>collapse</dt>
<dd><p>a boolean indicating whether to collapse multiple character combinations into only the observed states. For example, if true a two character dataset contained (0,0), (1,0), and (1,1), this would be collapsed into 1,2,3. However, if set to false it would 1,2,4. In combination with a custom rate matrix this allows for the estimation of transitions between the unobserved character combination. The default is <code>TRUE</code></p></dd>
</dl><p>.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>This is a stand alone function for computing the marginal, joint, or scaled likelihoods of internal nodes for a given set of transition rates. Like all other functions contained in corHMM, the tree does not have to be bifurcating in order for analyses to be carried out. IMPORTANT: If the corDISC, corHMM, and rayDISC functions are used they automatically provide a tree with the likeliest states as internal node labels. This function is intended for circumstances where the user would like to reconstruct states based on rates estimated elsewhere (e.g. BayesTraits, Mesquite, <code>ape</code>).</p>
<p>The algorithm based on Pupko et al. (2000, 2002) is used to calculate the <code>joint</code> estimates of ancestral states. The <code>marginal</code> method was originally implemented based on a description of an algorithm by Yang (2006). The basic idea is that the tree is rerooted on each internal node, with the marginal likelihood being the probabilities of observing the tips states given that the focal node is the root. However, this takes a ton of time as the number of nodes increase. But, importantly, this does not work easily when the model contains asymmetric rates. Here, we use the same dynamic programming algorithm as Mesquite (Maddison and Maddison, 2011), which is time linear with the number of species and calculates the marginal probability at a node using an additional up and down pass of the tree.  If <code>scaled</code>, the function uses the same algorithm from ace(). Note that the <code>scaled</code> method of ace() is simply the conditional likelihoods of observing everything at or above the focal node and these should NOT be used for ancestral state estimation.</p>
<p>The user can fix the root state probabilities by supplying a vector to <code>root.p</code>. For example, in the two trait case, if the hypothesis is that the root is 00, then the root vector would be <code>root.p=c(1,0,0,0)</code> for state combinations 00, 01, 10, and 11, respectively. If analyzing a binary or multistate character, the order of root.p is the same order as the traits -- e.g., for states 1, 2, 3, a <code>root.p=c(0,1,0)</code> would fix the root to be in state 2. If the user supplies the flag <code>root.p</code>=“yang”, then the estimated transition rates are used to set the weights at the root (see pg. 124 Yang 2006), whereas specifying <code>root.p</code>=“maddfitz” employs the same procedure described by Maddison et al. (2007) and FitzJohn et al. (2009). Note that the default <code>root.p=NULL</code> assumes equal weighting among all possible states.</p>
<p>Setting get.likelihood=TRUE will provide the user the joint likelihood of the rates and states.</p>
    </div>
    <div id="value">
    <h2>Value</h2>
    <dl><dt>$lik.tip.states</dt>
<dd><p>A matrix of the reconstructed tip values. If the number of rate.cats is greater than 2 then the probability that each observed state is in a particular hidden state is given.</p></dd>

<dt>$lik.anc.states</dt>
<dd><p>For <code>joint</code>, a vector of likeliest states at internal nodes and tips. For either <code>marginal</code> or <code>$scaled</code>, a matrix of the probabilities of each state for each internal node are returned.</p></dd>

<dt>$info.anc.states</dt>
<dd><p>A vector containing the amount of information (in bits) that the tip states and model gives to each node. See Boyko and Beaulieu (2021).</p></dd>

</dl></div>
    <div id="references">
    <h2>References</h2>
    <p>FitzJohn, R.G., W.P. Maddison, and S.P. Otto. 2009. Estimating trait-dependent speciation and extinction rates from incompletely resolved phylogenies. Systematic Biology 58:595-611.</p>
<p>Maddison, W.P. and D.R. Maddison. 2011. Mesquite: a modular system for evolutionary analysis. Version 2.75 http://mesquiteproject.org</p>
<p>Pupko, T., I. Pe'er, R. Shamir, and D. Graur. 2000. A fast algorithm for joint reconstruction of ancestral amino-acid sequences. Molecular Biology and Evolution 17:890-896.</p>
<p>Pupko, T., I. Pe'er, D. Graur, M. Hasegawa, and N Friedman N. 2002. A branch-and-bound algorithm for the inference of ancestral amino-acid sequences when the replacement rate varies among sites: application to the evolution of five gene families. Bioinformatics 18:1116-1123.</p>
<p>Yang, Z. 2006. Computational Molecular Evolution. London:Oxford.</p>
<p>Boyko, J. D., and J. M. Beaulieu. 2021. Generalized hidden Markov models for phylogenetic comparative datasets. Methods in Ecology and Evolution 12:468-478.</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Jeremy M. Beaulieu and Jeffrey C. Oliver</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># \donttest{</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">primates</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">phy</span> <span class="op">&lt;-</span> <span class="fu">multi2di</span><span class="op">(</span><span class="va">primates</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">data</span> <span class="op">&lt;-</span> <span class="va">primates</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="va">MK_3state</span> <span class="op">&lt;-</span> <span class="fu"><a href="corHMM.html">corHMM</a></span><span class="op">(</span>phy <span class="op">=</span> <span class="va">phy</span>, data <span class="op">=</span> <span class="va">data</span>, rate.cat <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Branch lengths of 0 detected. Adding 1e-5 to these branches.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> State distribution in data:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> States:	1	2	3	</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Counts:	29	10	21	</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Beginning thorough optimization search -- performing 0 random restarts </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Finished. Inferring ancestral states using marginal reconstruction. </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># # one way to get the parameters from your corHMM object in the correct order</span></span></span>
<span class="r-in"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">MK_3state</span><span class="op">$</span><span class="va">index.mat</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> </span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/stats/na.fail.html" class="external-link">na.omit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">MK_3state</span><span class="op">$</span><span class="va">solution</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/stats/na.fail.html" class="external-link">na.omit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">MK_3state</span><span class="op">$</span><span class="va">index.mat</span><span class="op">)</span> <span class="op">==</span> <span class="va">x</span><span class="op">)</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># using custom params</span></span></span>
<span class="r-in"><span><span class="va">states_1</span> <span class="op">&lt;-</span> <span class="fu">ancRECON</span><span class="op">(</span>phy <span class="op">=</span> <span class="va">phy</span>, data <span class="op">=</span> <span class="va">MK_3state</span><span class="op">$</span><span class="va">data</span>, p <span class="op">=</span> <span class="va">p</span>, method <span class="op">=</span> <span class="st">"marginal"</span>, </span></span>
<span class="r-in"><span>  <span class="va">rate.cat</span> <span class="op">&lt;-</span> <span class="va">MK_3state</span><span class="op">$</span><span class="va">rate.cat</span>, ntraits <span class="op">=</span> <span class="cn">NULL</span>, rate.mat <span class="op">=</span> <span class="va">MK_3state</span><span class="op">$</span><span class="va">index.mat</span>, </span></span>
<span class="r-in"><span>  root.p <span class="op">=</span> <span class="va">MK_3state</span><span class="op">$</span><span class="va">root.p</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Jeremy Beaulieu, Brian O’Meara, Jeffrey Oliver, James Boyko.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

